File: examples/rust/src/mutex.rs
// struct SinglyLinkedList {
//     value: i32,
//     child: Mutex<Arc<SinglyLinkedList>>,
// }

// fn delete(root: Mutex<SinglyLinkedList>, value: i32) -> {

// }

#[cfg(test)]
mod tests {
    use std::sync::{Arc, Mutex};
    use std::thread;

    #[test]
    fn test_mutex_count_across_threads() {
        let counter_start = 0;
        let num_threads = 5;

        // create a shared Mutex for the counter
        let counter = Arc::new(Mutex::new(counter_start));

        let mut thread_handles = vec![];

        for i in 0..num_threads {
            let counter = Arc::clone(&counter);

            // Spawn a thread that increments the counter
            let thread_handle = thread::spawn(move || {
                // entering "critical section"
                let mut counter = counter.lock().expect("unable to acquire mutex");

                // simulate some work
                thread::sleep(std::time::Duration::from_millis(i));

                // Increment the counter
                *counter += 1;
                println!("Thread {} incremented the counter to {}", i, *counter);
            });

            thread_handles.push(thread_handle);
        }

        // wait for all threads to finish
        for handle in thread_handles {
            handle.join().unwrap();
        }

        // access the final counter value
        let final_counter = *counter.lock().expect("unable to acquire mutex");
        println!("Final counter value: {}", final_counter);

        // test assertion
        assert_eq!(final_counter, counter_start + num_threads)
    }
}

File: examples/c++/tests/mutex.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <gtest/gtest.h>

std::mutex mtx;     // Mutex to protect access to shared data
int sharedData = 0; // Shared data

const int numThreads = 5;

void incrementData(int id)
{
    for (int i = 0; i < 5; ++i)
    {
        // Lock the mutex to protect the critical section
        std::unique_lock<std::mutex> lock(mtx);

        // Simulate some work
        std::this_thread::sleep_for(std::chrono::milliseconds(id));

        // Increment the shared data
        ++sharedData;
        std::cout << "Thread " << id << " incremented sharedData to " << sharedData << std::endl;

        // Unlock the mutex when done
        lock.unlock();

        // Continue with other work outside the critical section
        std::this_thread::sleep_for(std::chrono::milliseconds(id));
    }
}

// Demonstrate some basic assertions.
TEST(HelloTest, BasicAssertions)
{
    // Expect two strings not to be equal.
    EXPECT_STRNE("hello", "world");
    // Expect equality.
    EXPECT_EQ(7 * 6, 42);
}

// int main() {
//     std::thread threads[numThreads];

//     // Create threads that increment the shared data
//     for (int i = 0; i < numThreads; ++i) {
//         threads[i] = std::thread(incrementData, i);
//     }

//     // Wait for all threads to finish
//     for (int i = 0; i < numThreads; ++i) {
//         threads[i].join();
//     }

//     // Print the final value of the shared data
//     std::cout << "Final sharedData value: " << sharedData << std::endl;

//     return 0;
// }